### PR TITLE
Initial for feedback for handle async requests.

### PR DESCRIPTION
**Pull request by cwallsfdc**
Seeking feedback.  Once approved, I'll update whatever else is needed and add tests.

Note `###` loglines (I'll eventually remove "###" prefix).
```
2020-06-29T18:12:14 New invocation started
2020-06-29T18:12:14 Streaming pipeline initialized
2020-06-29T18:12:14 Ready to process data
2020-06-29T18:12:14 ### Received initial async request
2020-06-29T18:12:14 ### Forwarded async request
2020-06-29T18:12:14 ### Return initial async request
2020-06-29T18:12:14 New invocation started
2020-06-29T18:12:14 Streaming pipeline initialized
2020-06-29T18:12:14 Ready to process data
2020-06-29T18:12:14 DEBUG flag is unset
2020-06-29T18:12:14 Invoking AsyncApexFunction with payload {}
```

Follow PR will be capturing function response and posting to Core API.

### CODE REVIEW

File: middleware/index.ts [hunk 1]
```diff
@@ -282,6 +284,58 @@ function errorMessage(error: Error): any {
...
 
+function isInitialAsyncRequest(payload: any, headers: any): boolean {
+    return payload.type 
+        && payload.type.startsWith('com.salesforce.function.async')
+        && !headers[ASYNC_HEADER];
+}
+
+// Invoke given function again to fulfill async request; response is not handled
+async function invokeAsyncFn(logger: Logger, payload: any, headers: any): Promise<void> {
+    let hostUrl = headers['Host'] || payload?.data?.sfContext?.resource;
```
Comment (jabrown85) [inline middleware/index.ts:295]: Is there a reason we prefer the host over the context resource?

File: middleware/index.ts [hunk 2]
```diff
@@ -282,6 +284,58 @@ function errorMessage(error: Error): any {
...
+    if (!hostUrl) {
+        throw new Error('Unable to determine host');
```
Comment (jabrown85) [inline middleware/index.ts:297]:
```suggestion
        throw new Error('Unable to determine host for async invocation');
```

File: middleware/index.ts [hunk 3]
```diff
@@ -282,6 +284,58 @@ function errorMessage(error: Error): any {
...
+    }
+
+    headers[ASYNC_HEADER] = 'true';
+    const parsedUrl = require('url').parse(hostUrl);
```
Comment (jabrown85) [inline middleware/index.ts:301]: I'd suggest moving `require(`url`)` at the top of the file. A single instance of the middleware is held onto for the duration of the pod's life. `require` can be expensive sometimes.

File: middleware/index.ts [hunk 4]
```diff
@@ -282,6 +284,58 @@ function errorMessage(error: Error): any {
...
+    const https = parsedUrl.protocol.startsWith('https');
+
+    const options = {
+        hostname: parsedUrl.hostname,
+        port: parsedUrl.port || (https ? 443 : 80),
+        method: 'POST',
+        path: '/',
+        headers
```
Comment (jabrown85) [inline middleware/index.ts:309]:
```suggestion
        {...headers, ASYNC_HEADER: true}
```

File: middleware/index.ts [hunk 5]
```diff
@@ -294,6 +348,15 @@ export default async function systemFn(message: any): Promise<any> {
     const requestId = headers['Ce-Id'] || headers['X-Request-Id'];
     const requestLogger = createLogger(requestId);
     try {
+        // If initial async request, invoke function again and release request
+        if (isInitialAsyncRequest(payload, headers)) {
+            requestLogger.info('### Received initial async request');
+            await invokeAsyncFn(requestLogger, payload, headers);
+            requestLogger.info('### Return initial async request');
+            // TODO: How to send 202?
```
Comment (jabrown85) [inline middleware/index.ts:356]:
```suggestion
            Message.builder()
              .addHeader('content-type', 'application/json')
              .addHeader('x-http-status', '202')
              .payload({})
              .build();
```

File: middleware/index.ts [hunk 6]
```diff
@@ -282,6 +284,58 @@ function errorMessage(error: Error): any {
...
+        method: 'POST',
+        path: '/',
+        headers
+    };
+
+    // Invoke function and ignore response
+    const lib = require(https ? 'https' : 'http');
+    await new Promise((resolve, reject) => {
+        const req = lib.request(options);
+        req.on('error', (err) => {
```
Comment (jabrown85) [inline middleware/index.ts:316]: I can't recall if we said we would try to write back to core here as well or not, but I guess this is where we would do it once we have the core writeback written for the happy path.

File: middleware/index.ts (hunk 1 above)
Comment (cwallsfdc) [reply middleware/index.ts:295]: Good question.  Making use of a standard header.  I believe Core sends.  `:invoke` needs updating.  If a client doesn't provide, we'll fallback to `resource`, although `sfContext` is internal.  Seems like there should be a place to store as top-level CloudEvent attr.

File: middleware/index.ts (hunk 3 above)
Comment (cwallsfdc) [reply middleware/index.ts:301]: `require` caches, right?  No problem to move to top.  Can do same for `http/s` lib.

File: middleware/index.ts (hunk 6 above)
Comment (cwallsfdc) [reply middleware/index.ts:316]: Yes, good call.  I'll add.

File: middleware/index.ts (hunk 3 above)
Comment (jabrown85) [reply middleware/index.ts:301]: Yeah, it does cache. I guess it really matters where we want to pay for the initial require cost.

**Pull request closed without merging**